docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+
+We can also add a new field to a constructor, similarly to templates -
+for example, add a ``sideLen`` field to our ``Polygon`` constructor, to specify
+the lengths of the sides of the polygon.
+
+.. code:: daml
+
+  data Shape
+    = Circle
+    | Polygon
+        { sides : Int
+        , sideLen : Numeric 10 -- New field
+        }
+    | Line { len : Numeric 10 }
+
+Building this fails because the new ``sideLen`` field is non-optional.
+
+.. code:: bash
+
+  > cd v2/my-pkg
+  > daml build
+  ...
+  error type checking data type Main.Shape:
+    The upgraded variant constructor Polygon from variant Shape has added a field.
+  ERROR: Creation of DAR file failed.
+
+Making the new ``sideLen`` field optional fixes the error:
+
+.. code:: daml
+
+  ...
+        , sideLen : Optional (Numeric 10) -- New field
+  ...
+
+.. code:: bash
+
+  > cd v2/my-pkg
+  > daml build
+  ...
+  Created .daml/dist/my-pkg-2.0.0.dar
+
+Limitations in Upgrading Variants
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Upgrading variants has some limitations - because we have defined the
+``Circle`` constructor without a field in curly braces, it cannot be
+upgraded with new fields.
+
+.. code:: daml
+
+  ...
+    -- Add a field where no fields existed
+    = Circle { radius : Optional (Numeric 10) }
+  ...
+
+.. code:: bash
+
+  > cd v2/my-pkg
+  > daml build
+  ...
+  error type checking data type Main.Shape:
+    The upgraded data type Shape has changed the type of a variant.
+  ERROR: Creation of DAR file failed.
+
+The same applies to variants with unnamed fields. If the v1 definition
+of the ``Line`` constructor were as follows, it would also not be able to
+upgrade:
+
+.. code:: daml
+
+  ...
+    | Line (Numeric 10)
+
+In general, in order to enable future upgrades, it is strongly
+recommended that all constructors use named fields, and that all
+constructors have at least one field. If a constructor has no fields in
+an initial v1 package, one can assign a dummy field.
+
+For example, the correct way to write the v1 ``Circle`` constructor would be
+as follows:
+
+.. code:: daml
+
+  ...
+    = Circle { dummy : () }
+  ...
+
+The subsequent v2 upgrade would succeed:
+
+.. code:: daml
+
+  ...
+    = Circle { dummy : (), radius : Optional (Numeric 10) }
+  ...
+
+Nested Datatypes
+~~~~~~~~~~~~~~~~
+
+If a data type, choice, or template has a field which refers to another
+data type, the larger data type will be upgradeable if the field’s data
+type is upgradeable as well.
+
+For example, given the data type ``A`` with a field referring to data type
+``B``,
+
+.. code:: daml
+
+  data A = A { b : B }
+  data B = B { field : Text }
+
+If modifications made to ``B`` are valid for SCU, then ``A`` will also be valid.
+
+Dependencies
+~~~~~~~~~~~~
+
+Package authors may upgrade the dependencies of a package as well as the
+package itself. A new version of a package may add new dependencies, and
+must have all the dependencies of the old version. Dependencies of the
```

Comment:
We'll add a note about this

Reply:
We ideally want to avoid details of utility packages in the main doc, as for most people, they're simply an implementation detail of how we made stdlib work. Instead, I've simply linked to their definition in the reference.